README.md
# http2amqp

[![Build Status](https://app.travis-ci.com/aleasoluciones/http2amqp.svg?branch=master)](https://app.travis-ci.com/github/aleasoluciones/http2amqp)
[![GoDoc](https://godoc.org/github.com/aleasoluciones/http2amqp?status.png)](http://godoc.org/github.com/aleasoluciones/http2amqp)
[![License](https://img.shields.io/github/license/aleasoluciones/http2amqp)](https://github.com/aleasoluciones/http2amqp/blob/master/LICENSE)

HTTP interface to AMPQ.

## Features

- It publishes an AMQP request message for each HTTP request received. Then it waits for the corresponding AMQP response message in order to answer to the original HTTP request.
- The topic of the message that it publishes comes from the method and the path of the HTTP request, replacing slashes (/) with dots (.).
  - e.g.: `GET http://localhost:18080/net/test` → `get.net.test`
- The topic of the response messages to close the connection is `queries.response`. These response messages must carry the same ID as the request they intend to close.
- The default TTL for the published messages is 1000 ms. After that, the connection will be closed with a timeout if a response did not arrive.
- The exchange used by default is 'events'.


## Build

You need a Go runtime installed in your system which supports [modules](https://tip.golang.org/doc/go1.16#modules). A nice way to have multiple Go versions and switch easily between them is the [g](https://github.com/stefanmaric/g) application.

A [Makefile](Makefile) is available, so you only need to run:

```sh
make build
```

## Running tests

Make sure that you have built the binaries from the previous section, because the tests will run the `echoservice` binary. 

Load environment variables to set the BROKER_URI environment variable.

```sh
source dev/env_develop
```

Start a RabbitMQ service with default configuration (specified in [`/dev/env_develop`](/dev/env_develop)).

```sh
make start_dependencies
```

Run tests. They will only work if the RabbitMQ container is up.

```sh
make test
```

## Building docker images

The Jenkins [pipeline](Jenkinsfile) will generate two containers, which we can also build locally. One for compiling the app (http2amqp-builder), and the other to deploy its binary (http2amqp). This avoids the need of having Go installed in the host system.  Take a look at the [Dockerfile](Dockerfile) for more details.

```sh
make build_images
```

Once built, if we want to run the tests inside the builder image, we can do the following (remember to have the environment variables loaded and the RabbitMQ container up too):

```sh
docker run --rm -it --net=host aleasoluciones/http2amqp-builder:GIT_REV integration-tests
```

And to run http2amqp from inside the container:

```sh
docker run --rm -it --net=host -e BROKER_URI=amqp://guest:guest@localhost:5666/ aleasoluciones/http2amqp:GIT_REV
```

## Usage

```sh
$ ./http2amqp -help
Usage of ./http2amqp:
  -address string
    	HTTP listen IP address (default "0.0.0.0")
  -brokeruri string
    	AMQP broker connection URI (default "amqp://guest:guest@localhost/")
  -exchange string
    	AMQP broker exchange name (default "events")
  -port string
    	HTTP listen port (default "18080")
  -timeout int
    	AMQP broker queries timeout in milliseconds (default 1000)
  -verbose
    	Verbose mode, enable logging
```

## Execution example

Make sure that the environment variables are loaded before executing each command. Also that you have a RabbitMQ container up and running with those parameters.

Start the htt2amqp server in a terminal.

```sh
./http2amqp -verbose
```

Start the echo service in another terminal. This will publish a response event each time a request event arrives, with the same ID and payload (ping pong).

```sh
./echoservice
```

Make a request with a payload. The echo service will answer back if running, otherwise it will timeout.

```
curl -X GET http://localhost:18080/net/test -d 'hello world'
```

You can specify the timeout (in milliseconds) as a query param.

```
curl -X GET http://localhost:18080/net/test?timeout=200 -d 'hello world'
```

## TODO

- Test timeout parameter for each request.
- Implement delay parameter for echo server to allow tests timeouts.

## Notes

* We compile in the Dockerfile with CGO_ENABLED=0 because in the scratch image there are not some C libraries it needs. So we compile the application with them in order to run it properly. See https://go.dev/blog/cgo for more info.


CHANGELOG.md
## [1.0.0] - 2021-12-17
* Migrate to Go modules, work with Go 1.17
* Add Jenkins pipeline
* Update dependencies
* Add docker-compose for running a RabbitMQ in dev
* Add environment variables
* Clean Makefile
* Improve README

## [0.4.8] - 2019-02-22
* Port is parameterized

## [0.4.7] - 2018-10-29
* Internal: Update go version 1.10 and 1.11
* Internal: Update godep version
* Internal: Update Makefile and dependencies
* Internal: Update go version for docker image

## [0.4.6] - 2018-10-25
* Systems: Use network host mode

## [0.4.5] - 2018-01-30
* Internal: Added godep for dependency management
            - Updated Makefile
            - Excluded vendor directory from git
            - Updated .travis
            - Updated dependencies

## [0.4.4] - 2017-12-15
* Internal: Removed go version 1.6 and 1.7. Added 1.8.x, 1.9.x and master versions

## [0.4.3] - 2016-10-24

* Use manual prefix "docker/" for docker_log_tag.It's necessary for docker v1.12
* Update go version to 1.6.3
* Travis configuration uses 1.6 1.7 and tip
* golint is not supported in go version 1.5
* Remove update_deps doesnt work correctly :-(

## [0.4.2] - 2016-04-20

* Update deprecated docker compose configuration syslog-tag

## [0.4.2] - 2016-04-19

* Internal: Unify Makefile and travis.yml

## [0.4.2] - 2016-01-11

* Publish queries with the timeout value as amqp message ttl so messages are expired after timeout

## [0.4.1] - 2015-12-30

* Bugfix: ignore messages that can not be deserialized.

## [0.4.0] - 2015-12-02

* return 408 error code when timeout happens

## [0.4.0] - 2015-08-11

* Support a query string argument to especify a request timeout in milliseconds (e.g. curl http://localhost/foo?timeout=1500)

## [0.3.0] - 2015-06-17

* Support multiple methods
* simplified code
* API changes

## [0.2.0] - 2015-02-24

* Bugfix. Memory leak in the map that stores pending responses.
* Improved logging to show query id, topic and criteria.

## [0.1.1] - 2015-02-23

* Nothing. Just released to test deployment

## [0.1.0] - 2015-02-23

* Initial release


httpserver/http2amqp.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	_ "net/http/pprof"
	"os"
	"strings"
	"time"

	alealog "github.com/aleasoluciones/goaleasoluciones/log"
	"github.com/aleasoluciones/http2amqp"
)

const defaultTimeoutMs = 1000
const defaultExchange = "events"
const defaultHTTPPort = "18080"
const defaultHTTPAddress = "0.0.0.0"
const defaultBrokerURI = "amqp://guest:guest@localhost/"

func main() {
	alealog.Init()
	alealog.DisableLogging()
	defaultVerbose := verboseMode()

	verbose := flag.Bool("verbose", defaultVerbose, "Verbose mode, enable logging")
	amqpuri := flag.String("amqpuri", envBrokerURI(), "AMQP connection uri")
	HTTPAddress := flag.String("address", envHTTPAddress(), "HTTP Listen address")
	HTTPPort := flag.String("port", envHTTPPort(), "HTTP Listen port")
	exchange := flag.String("exchange", defaultExchange, "AMQP exchange name")
	timeout := flag.Int("timeout", defaultTimeoutMs, "AMQP Queries timeout in milliseconds")
	flag.Parse()

	if *verbose {
		alealog.EnableLogging()
		log.Println("[http2amqp] verbose mode enabled")
	}

	service := http2amqp.NewService(*amqpuri, *exchange, time.Duration(*timeout)*time.Millisecond)

	http.HandleFunc("/", http2amqp.NewHTTPServerFunc(service))
	addressAndPort := fmt.Sprintf("%s:%s", *HTTPAddress, *HTTPPort)
	log.Println("[http2amqp] Starting HTTP server at ", addressAndPort)
	http.ListenAndServe(addressAndPort, nil)
}

func envHTTPPort() string {
	port := os.Getenv("PORT")

	if len(port) == 0 {
		port = defaultHTTPPort
	}

	return port
}

func envHTTPAddress() string {
	address := os.Getenv("ADDRESS")

	if len(address) == 0 {
		address = defaultHTTPAddress
	}

	return address
}

func envBrokerURI() string {
	brokerURI := os.Getenv("BROKER_URI")

	if len(brokerURI) == 0 {
		brokerURI = defaultBrokerURI
	}

	return brokerURI
}

func isTrue(value string) bool {
	value = strings.ToLower(value)
	trulies := []string{"1", "y", "yes", "on"}

	for _, truly := range trulies {
		if value == truly {
			return true
		}
	}
	return false
}

func verboseMode() bool {
	verbose := os.Getenv("HTTP2AMQP_VERBOSE")
	return isTrue(verbose)
}
